js-peer/src/context/chat-ctx.tsx

```diff
@@ -119,7 +172,46 @@ export const ChatProvider = ({ children }: any) => {
   }
 
   useEffect(() => {
-    libp2p.services.pubsub.addEventListener('message', messageCB)
+    const handleDirectMessage = async (evt: CustomEvent<DirectMessageEvent>) => {
+      const peerId = evt.detail.connection.remotePeer.toString()
+
+      if (evt.detail.type !== mimeTextPlain) {
+        throw new Error(`unexpected message type: ${evt.detail.type}`)
+      }
+
+      const message: ChatMessage = {
+        msg: evt.detail.content,
+        from: 'other',
+        read: false,
+        msgId: crypto.randomUUID(),
+        fileObjectUrl: undefined,
+        peerId: peerId,
+        receivedAt: Date.now(),
+      }
+
+      const updatedMessages = directMessages[peerId]
+        ? [...directMessages[peerId], message]
+        : [message]
+
+      setDirectMessages({
+        ...directMessages,
+        [peerId]: updatedMessages,
+      })
+    }
+
+    const eventListener = (evt: Event) => {
+      handleDirectMessage(evt as CustomEvent<DirectMessageEvent>);
+    };
+
+    document.addEventListener(directMessageEvent, eventListener)
```

Comment:
What do you think about [creating a custom service](https://github.com/libp2p/js-libp2p/blob/main/doc/SERVICES.md) with its own eventTarget for these messages to avoid relying on the global `document`? 

Might be a good place to encapsulate the protocol logic.

Reply:
Nice - hadn't seen that doc, will refactor

Reply:
@2color implemented via a custom service - much cleaner! thanks for the suggestion

Reply:
Fantastic. Will take a look this week as soon as I have a bit of time.

Reply:
Looking neat!